src/functional/ffront/foast_pretty_printer.py

```diff
@@ -0,0 +1,195 @@
+    foast.Call: 17,
+    foast.Attribute: 17,
+    # Await expression
+    #  not used
+    # Exponentiation
+    (foast.BinOp, "power"): 16,
+    # Positive, negative, bitwise NOT
+    (foast.UnaryOp, "plus"): 14,
+    (foast.UnaryOp, "minus"): 14,
+    (foast.UnaryOp, "not_"): 14,
+    # Multiplication, matrix multiplication, division, floor division, remainder
+    (foast.BinOp, "multiplies"): 13,
+    (foast.BinOp, "divides"): 13,
+    (foast.BinOp, "floordiv"): 13,
+    # (foast.BinOp, "remainder"): 13,
+    # Addition and subtraction
+    (foast.BinOp, "plus"): 12,
+    (foast.BinOp, "minus"): 12,
+    # Shifts
+    #  not used
+    # Bitwise AND
+    #  not used
+    # Bitwise XOR
+    #  not used
+    # Bitwise OR
+    #  not used
+    # Comparisons, including membership tests and identity tests
+    (foast.Compare, "less"): 7,
+    (foast.Compare, "less_equal"): 7,
+    (foast.Compare, "greater"): 7,
+    (foast.Compare, "greater_equal"): 7,
+    (foast.Compare, "not_eq"): 7,
+    (foast.Compare, "eq"): 7,
+    # Boolean NOT
+    (foast.UnaryOp, "not_"): 6,
+    # Boolean AND
+    (foast.BinOp, "and_"): 5,
+    # Boolean OR
+    (foast.BinOp, "or_"): 4,
+    # Conditional expression
+    foast.TernaryExpr: 3,
+    # Lambda expression
+    #  not used
+    # Assignment expression
+    #  not used
+}
+
+RIGHT_GROUPING = {(foast.BinOp, "power"), foast.TernaryExpr}
+
+
+class Group(enum.IntEnum):
+    LEFT = 0
+    RIGHT = 1
+
+
+def _property_identifier(node: foast.LocatedNode) -> PropertyIdentifier:
+    if isinstance(node, foast.BinOp):
+        return (foast.BinOp, node.op)
+    elif isinstance(node, foast.UnaryOp):
+        return (foast.UnaryOp, node.op)
+    elif isinstance(node, foast.Compare):
+        return (foast.Compare, node.op)
+
+    return type(node)
+
+
+class PrettyPrinter(codegen.TemplatedGenerator):
+    Symbol = as_fmt("{id}")
+
+    Name = as_fmt("{id}")
+
+    Constant = as_fmt("{value}")
+
+    Subscript = as_fmt("{value}[{index}]")
+
+    Attribute = as_fmt("{value}.{attr}")
+
+    TupleExpr = as_fmt("({', '.join(elts)}{',' if len(elts)==1 else ''})")
+
+    UnaryOp = as_fmt("{op}{operand}")
+
+    def visit_BinOp(self, node: foast.BinOp, **kwargs) -> str:
+        left = self._parenthesize(node.left, node, Group.LEFT)
+        right = self._parenthesize(node.right, node, Group.RIGHT)
+        return f"{left} {node.op} {right}"
+
+    def visit_Compare(self, node: foast.Compare, **kwargs) -> str:
+        left = self._parenthesize(node.left, node, Group.LEFT)
+        right = self._parenthesize(node.right, node, Group.RIGHT)
+        return f"{left} {node.op} {right}"
+
+    def visit_TernaryExpr(self, node: foast.TernaryExpr, **kwargs) -> str:
+        cond = self.visit(node.condition)
+        true_expr = self._parenthesize(node.true_expr, node, Group.LEFT)
+        false_expr = self._parenthesize(node.false_expr, node, Group.RIGHT)
+        return f"{true_expr} if {cond} else {false_expr}"
```

Comment:
can you move these function after all as_fmt calls, i.e. after the Return definition